### ISSUE TITLE
Scaling to high parameter-of-interest dimensions

### ISSUE THREAD

**Issue by rafaelorozco** [lee-group-cmu/lf2i#3]
Hello, 

I attended the talks about these methods at UQ Trieste and was excited to implement this on our conditional normalizing flow SBI projects. But it seems that there are many operations that are done at a per parameter dimension basis. Does this mean that this algorithm is difficult to scale when the POI has a high dimension? Our applications are for subsurface earth imaging so require scaling to POI that are image volumes of 512^3.

Thank you for your attention

**Comment by lucamasserano** [comment]
Hi Rafael, 

Thanks for your interest in LF2I! When scaling to such a high POI dimensionality, I foresee two things would need to be adapted:
- **critical values estimation:** at the moment the default is to use gradient boosting, but I also have a simple feedforward neural network implemented that minimizes the quantile loss (see [here](https://github.com/lee-group-cmu/lf2i/blob/dev/src/lf2i/critical_values/nn_qr_algorithm.py)). In your case, given that parameters are images, none of these would work well right away. The way I would do it is to start from that neural network and expand it using some dedicated CNN/Transformer architecture that allows to exploit the right inductive bias. Estimating critical values directly from a 512^3 parameter space without learning any appropriate representation in the middle would be hopeless, I think.
- **construction of confidence sets**: the current implementation requires you to provide a grid of points over the POI space that is used to construct the confidence set, which is not ideal in your case and would likely cause memory errors once the test statistic is evaluated all-at-once on the grid. For your application, I would probably setup something that samples plausible POIs (images) and passes them one by one to the pipeline to construct a confidence set. 

These are the two first issues that come to my mind. Please let me know if they don't exactly answer your questions or if you need more information! I am happy to follow-up and help out. I am also in the process of implementing other functionalities for the library so I will try to take into account these very high-dimensional cases.

**Comment by rafaelorozco** [comment]
Hello thank you for the reply,

Congrats on the great work and great talk at the conference! Before we continue I have a fundamental question that would help guide my confusion:

in general, I am confused on exactly what the output is of your algorithm. With SBI + conditional normalizing flows we have a quantification of the uncertainty by looking at the variance of the posterior samples. I.e. an UQ image of the variance of the posterior samples. Would your algorithm return a new UQ image?  By the way, for now I am happy with a more calibrated UQ image. I am okay if we do not do the "diagnostics" part of your workflow.

**Comment by lucamasserano** [comment]
Hi Rafael, 

LF2I outputs a confidence set for your parameter of interest $\theta$. That is, a set $\mathcal{R}(X)$ such that $\mathbb{P}_{X|\theta}(\theta^* \in \mathcal{R}(X)) \approx 1-\alpha$. 

Intuitively, and specifically for your case, LF2I should output a set of “plausible” images that are guaranteed to contain the ”true” image (that generated your observation) with some probability. The size of this set determines the level of uncertainty around the POI: the smaller/tighter the set, the lower the uncertainty. If the parameter space is euclidean, one can compute this size by measuring the volume of the set. In your case, we would need to understand what is an appropriate measure of size of this set, or maybe look at worst/best images in this set, where worst/best depends on your use case.

Having a better idea of what is your specific application and data-generating process would help me phrasing this explanation in a context that is more familiar to your problem. 

As for the diagnostics, that's only an independent check of the coverage of the constructed confidence sets. The two points I mentioned in the first answer above relate directly to how we can construct these confidence sets.

**Comment by rafaelorozco** [comment]
Hello, 

Our application is full-waveform inversion. The data generation process is the forward wave equation, restricted to receivers at the surface while the parameter of interest is the wave velocity properties of the subsurface. The goal is to scale to POI in volumes thus try to avoid repeating the data generation process too many times since it is expensive to solve the wave PDE. We have access to around 1000 training examples of the POI so are willing to perform 1000 simulations and can perform SBI with neural posterior learning: https://arxiv.org/pdf/2401.06230.pdf
but we are very interested in improving the calibration of our UQ as given by the variation of the posterior samples. 

This part:
` I would probably setup something that samples plausible POIs (images) and passes them one by one to the pipeline to construct a confidence set.` 
seems difficult for us since we do not have an arbitrary amount of prior samples to work with. Instead we have a relatively small training dataset (around 1000). And I remember you saying in the talk that for technical reasons we prefer to not reuse the training dataset.


Thank you for the help! I owe you a coffee at the next conference!

**Comment by lucamasserano** [comment]
Hi Rafael, 

> This part:
 I would probably setup something that samples plausible POIs (images) and passes them one by one to the pipeline to construct a confidence set.

One thing you could do is to use the POIs in the training set as the evaluation grid. The question is if they approximately cover the support of your parameter space, because ideally you would want to construct confidence sets that take into consideration all plausible values of the POI. Using those you have in the training set would be a good start in my opinion.

> And I remember you saying in the talk that for technical reasons we prefer to not reuse the training dataset.

Yes, we haven't thoroughly investigated the effect of not doing that, but for reasons related to overfitting on the training set the coverage guarantee of the confidence regions we construct would hold if you split your data into training and calibration. This is similar to split conformal prediction, if you are familiar with it. This said, if you have very limited data you could try using overlapping sets and see what you get

> Thank you for the help! I owe you a coffee at the next conference!

No worries! Very happy to help, and thanks for your questions. Definitely let me know if you need more help. We could also organize a zoom call if you want to discuss some things more specifically. Hope to meet you at the next conference!

**Comment by rafaelorozco** [comment]
Great thank you for the back and forth. I think that this gives me enough food for though so that I can try some things out. If I have more specific questions I will come back but for now this issue has been resolved.
